README.md

```diff
@@ -1,34 +1,332 @@
-# Crate Name
+# vm-allocator crate
 
 ## Design
+vm-allocator is designed about resource and resource allocator being responsible
+for system resources allocation for virtual machine(VM).
 
-TODO: This section should have a high-level design of the crate.
+It provides different kinds of resources that might be used by VM, such as
+memory-maped I/O address space, port I/O address space, legacy IRQ numbers, MSI/MSI-X
+vectors, device instance id, etc. All these resources should be implemented with
+allocation and freeing mechanism in this crate, in order to make VMM easier to
+construct.
 
-Some questions that might help in writing this section:
-- What is the purpose of this crate?
-- What are the main components of the crate? How do they interact which each
-  other?
+Main components are listed below.
+- A `Resource` trait representing for all kinds of resources.
+```rust
+pub trait Resource {}
+```
+
+- A `ResourceSize` trait representing for all kinds of resource size.
+```rust
+pub trait ResourceSize {}
```

Comment:
Having a separate trait for the size of a resource is making the code complicated (for example in the `update` case where you need to send 4 parameters just to update one resource). Isn't it possible to specify that `Resource` should implement `std::marker::Sized`?